libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +175,155 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
+
+    const size_t outVertexCount = triangleCount * 3;
+    float3* outPositions = new float3[outVertexCount];
+    float2* outUvs = uvs ? new float2[outVertexCount] : nullptr;
+    quatf* quats = new quatf[outVertexCount];
+
+    const size_t outTriangleCount = triangleCount;
+    uint3* outTriangles = new uint3[outTriangleCount];
+
+    size_t vindex = 0;
+    for (size_t tindex = 0; tindex < triangleCount; ++tindex) {
+        uint3 tri = isTriangle16 ?
+                uint3(pointerAdd<ushort3, uint8_t>(triangles, tindex, tstride)) :
+                pointerAdd<uint3, uint8_t>(triangles, tindex, tstride);
+
+        const float3 pa = pointerAdd<float3, float3>(positions, tri.x, pstride);
+        const float3 pb = pointerAdd<float3, float3>(positions, tri.y, pstride);
+        const float3 pc = pointerAdd<float3, float3>(positions, tri.z, pstride);
+
+        uint32_t i0 = vindex++, i1 = vindex++, i2 = vindex++;
+        outTriangles[tindex] = uint3{i0, i1, i2};
+
+        outPositions[i0] = pa;
+        outPositions[i1] = pb;
+        outPositions[i2] = pc;
+
+        const float3 n = normalize(cross(pc - pb, pa - pb));
+        float3 t, b;
+        frisvadKernel(n, t, b);
+
+        const quatf tspace = mat3f::packTangentFrame({t, b, n});
+        quats[i0] = tspace;
+        quats[i1] = tspace;
+        quats[i2] = tspace;
+
+        if (outUvs) {
+            outUvs[i0] = pointerAdd<float2, float2>(uvs, tri.x, uvstride);
+            outUvs[i1] = pointerAdd<float2, float2>(uvs, tri.y, uvstride);
+            outUvs[i2] = pointerAdd<float2, float2>(uvs, tri.z, uvstride);
+        }
+    }
+
+    output->tangentSpace = quats;
+    output->positions = outPositions;
+    output->vertexCount = outVertexCount;
+    output->uvs = outUvs;
+    output->triangles32 = outTriangles;
+    output->triangleCount = outTriangleCount;
+}
+
+template<typename DataType, typename InputType>
+inline void cleanOutputPointer(DataType*& ptr, InputType inputPtr) {
+    if (ptr && ptr != (const DataType*) inputPtr) {
+        delete[] ptr;
+    }
+    ptr = nullptr;
+}
+
+template <typename DataType>
+inline void takeStride(DataType*& out, size_t stride) {
+    out = (DataType*) (((uint8_t*) out) + stride);
+}
+
+template <typename OutputType>
+using QuatConversionFunc = OutputType(*)(const quatf&);
+
+template<typename OutputType>
+inline void getQuatsImpl(OutputType* UTILS_RESTRICT out, const quatf* UTILS_RESTRICT tangentSpace,
+        size_t vertexCount, size_t stride, QuatConversionFunc<OutputType> conversion) {
+    stride = stride ? stride : sizeof(OutputType);
+    for (size_t i = 0; i < vertexCount; ++i) {
+        *out = conversion(tangentSpace[i]);
+        out = (OutputType*) (((uint8_t*) out) + stride);
+    }
+}
 
 } // anonymous namespace
 
 Builder::Builder() noexcept
         :mMesh(new TangentSpaceMesh()) {}
 
-Builder::~Builder() noexcept { delete mMesh; }
+Builder::~Builder() noexcept { }
```

Comment:
There is a leak here if the user never calls `build()`. `mMesh` won't be destroyed.